### ISSUE TITLE
`filter_vmap()` on `eqx.nn.*` Modules with `key` as a keyword-only argument

### ISSUE THREAD

**Issue by mthiboust** [patrick-kidger/equinox#405]
Thanks for the library!

If I understand correctly:
* Keyword arguments cannot be used with functions wrapped with `eqx.filter_vmap` 
* The `key` argument of all equinox Module methods is a keyword-only argument (see the [Dropout Module](https://github.com/patrick-kidger/equinox/blob/main/equinox/nn/_dropout.py#L60) for instance)

Does this mean that we cannot directly `filter_vmap` on an equinox Module `__call__` when using the `key` argument? If facing such a case, is it the best practice to wrap this call into an other positional-only-argument function? 

Example with `eqx.nn.Dropout`:

```python
import jax.numpy as jnp 
import jax.random as jrandom

dropout = eqx.nn.Dropout(0.2)

def dropout2(x, key):
    return dropout(x, key=key)

x = jnp.ones((3,5))
keys = jrandom.split(key, 3)

eqx.filter_vmap(dropout)(x, key=keys) # RuntimeError: keyword arguments cannot be used with functions wrapped with `filter_vmap`
eqx.filter_vmap(dropout)(x, keys) # TypeError: Dropout.__call__() takes 2 positional arguments but 3 were given
eqx.filter_vmap(dropout2)(x, keys) # OK
```

More generally, why did you choose to impose the `key` argument as a keyword-only argument in your Modules? Would you recommend us to always do the same or do you see edge cases where it is better no to follow this practice?

**Comment by patrick-kidger** [comment]
Right, this is an annoying interaction. I agree we should try to neaten this somehow.

So `filter_vmap` doesn't take keyword arguments for backward-compatibility reasons. There was a time when it did take such arguments but with different behaviour. Explicitly making it an error now ensured that end-users don't silently get something unexpected.

We could probably afford to relax that now though. I'd be happy to take a PR generalising things. (One point of caution: we need to handle things like `in_axes=dict(key=1)` correctly.)

---

As for why `key` is a keyword-only argument. There's no strong reason for this, I'm just generally a fan of keyword-only arguments. It helps to protect against accidentally passing arguments in the wrong position.

**Comment by mthiboust** [comment]
Thanks Patrick for the explanation
